File: java/leetcode_grind/Day427.java
package leetcode_grind;

public class Day427 {
    // https://leetcode.com/problems/insertion-sort-list/description/
    static class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    static class Solution1 {
        public ListNode insertionSortList(ListNode head) {
            var dummy = new ListNode(Integer.MIN_VALUE);
            var prev = dummy;

            var curr = head;
            while (curr != null) {
                var next = curr.next;
                curr.next = null;
                while (prev != null) {
                    if (prev.next == null || prev.next.val >= curr.val) {
                        curr.next = prev.next;
                        prev.next = curr;
                        break;
                    }
                    prev = prev.next;
                }
                curr = next;
                prev = dummy;
            }

            return dummy.next;
        }
    }
}

File: python/leetcode_grind/day_427.py
# https://www.geeksforgeeks.org/problems/insertion-sort-for-singly-linked-list/1
# Node Class
class Node:
    def __init__(self, data):   # data -> value stored in node
        self.data = data
        self.next = None

class Solution:
    def insertionSort(self, head):
        #code here
        #code here
        curr = head
        dummy = Node(-1000000007)
        prev = dummy

        while curr is not None:
            next = curr.next
            curr.next = None

            while prev is not None:
                if prev.next is None or prev.next.data >= curr.data:
                    curr.next = prev.next
                    prev.next = curr
                    break
                prev = prev.next

            curr = next
            prev = dummy

        return dummy.next